crates/runtime/src/instance/allocator.rs

```diff
@@ -0,0 +1,633 @@
+                    "memory out of bounds: data segment does not fit".into(),
+                )));
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn initialize_memories(
+    instance: &Instance,
+    initializers: &[MemoryInitializer],
+) -> Result<(), InstantiationError> {
+    for init in initializers {
+        let memory = instance.get_memory(init.memory_index);
+
+        let start = get_memory_init_start(init, instance);
+        if start
+            .checked_add(init.data.len())
+            .map_or(true, |end| end > memory.current_length)
+        {
+            return Err(InstantiationError::Trap(Trap::wasm(
+                ir::TrapCode::HeapOutOfBounds,
+            )));
+        }
+
+        unsafe {
+            let mem_slice = get_memory_slice(init, instance);
+            let end = start + init.data.len();
+            let to_init = &mut mem_slice[start..end];
+            to_init.copy_from_slice(&init.data);
+        }
+    }
+
+    Ok(())
+}
+
+fn check_init_bounds(instance: &Instance) -> Result<(), InstantiationError> {
+    check_table_init_bounds(instance)?;
+
+    match &instance.module.memory_initialization {
+        Some(MemoryInitialization::Paged { .. }) | None => {
+            // Bounds were checked at compile-time
+        }
+        Some(MemoryInitialization::OutOfBounds) => {
+            return Err(InstantiationError::Link(LinkError(
+                "memory out of bounds: data segment does not fit".into(),
+            )));
+        }
+        Some(MemoryInitialization::Segmented(initializers)) => {
+            check_memory_init_bounds(instance, initializers)?;
+        }
+    }
+
+    Ok(())
+}
+
+fn initialize_instance(
+    instance: &Instance,
+    is_bulk_memory: bool,
+) -> Result<(), InstantiationError> {
+    // If bulk memory is not enabled, bounds check the data and element segments before
+    // making any changes. With bulk memory enabled, initializers are processed
+    // in-order and side effects are observed up to the point of an out-of-bounds
+    // initializer, so the early checking is not desired.
+    if !is_bulk_memory {
+        check_init_bounds(instance)?;
+    }
+
+    // Initialize the tables
+    initialize_tables(instance)?;
+
+    // Initialize the memories
+    match &instance.module.memory_initialization {
+        Some(MemoryInitialization::Paged { page_size, map }) => {
+            for (index, pages) in map {
+                let memory = instance.memory(index);
+
+                for (page_index, page) in pages.iter().enumerate() {
+                    if let Some(data) = page {
+                        // Bounds checking should have occurred when the module was compiled
+                        // The data should always be page sized
+                        assert!((page_index * page_size) < memory.current_length);
+                        assert_eq!(data.len(), *page_size);
+
+                        unsafe {
+                            ptr::copy_nonoverlapping(
```

Comment:
I totally agree that using paged initialization might have a performance impact when not using uffd (which requires it so that we can quickly tell the kernel "copy this page exactly and wake the faulting thread").

I'd be totally fine with defaulting to segmented initialization always unless the uffd feature is enabled.  What's your thoughts on that?

Reply:
Yeah I think that sounds reasonable to me! We should have semi-comprehensive tests for uffd anyway which I think should cover our bases in ensuring all branches are covered effectively.